filebeat/channel/outlet.go

```diff
@@ -27,25 +27,32 @@ type outlet struct {
 	wg     eventCounter
 	client beat.Client
 	isOpen atomic.Bool
+	done   chan struct{}
 }
 
 func newOutlet(client beat.Client, wg eventCounter) *outlet {
 	o := &outlet{
 		wg:     wg,
 		client: client,
 		isOpen: atomic.MakeBool(true),
+		done:   make(chan struct{}),
 	}
 	return o
 }
 
 func (o *outlet) Close() error {
 	isOpen := o.isOpen.Swap(false)
 	if isOpen {
+		close(o.done)
 		return o.client.Close()
 	}
 	return nil
 }
 
+func (o *outlet) Done() <-chan struct{} {
```

Comment:
I am a bit on the fence with this change. I have read that it's an anti-pattern to return channels from functions. An alternative solution I can see is exposing `isDone` as a function and a timer to check if the `outlet` is running or not. The timer could fire every 5 minutes, so we could provide periodical "garbage" collection. But this might be too complex.

Reply:
I can see that returning channels can be problematic in some cases, but I think that it is fine for "done channels" that are used only for signaling (like in `Done()` in the [`context.Context`](https://golang.org/pkg/context/#Context) interface of the standard library).

Also for this case I think it'd make the code more complicated. As you mention we'd need to expose `isOpen` for the case of `outlet`, but we'd also need to add an equivalent variable to `subOutlet`. And in places where we had to wait for it to be closed we'd have to enclose the wait in some loop, like for `CloseOnSignal` here we'd have to do something like:
```
                for outlet.IsOpen() {
			select {
			case <-time.After(5 * time.Minute):
			case <-sig:
				outlet.Close()
                                return
			}
                }
```

And we'd still have unneeded goroutines alive during the time of the wait.